syntax = "proto3";

option java_package = "ms.domwillia.jvmemory.protobuf";
option java_outer_classname = "Message";

import "allocations.proto";
import "definitions.proto";
import "flow.proto";
import "access.proto";

enum MessageType {
    CLASS_DEF = 0;

    ALLOC_OBJECT = 1;
    ALLOC_ARRAY = 2;
    DEALLOC = 3;

    METHOD_ENTER = 4;
    METHOD_EXIT = 5;

    GETFIELD = 6;
    PUTFIELD_OBJECT = 7;
    PUTFIELD_PRIMITIVE = 8;

    STORE_OBJECT = 9;
    STORE_PRIMITIVE = 10;
    LOAD = 11;
}

message Variant {
    MessageType type = 1;
    int64 thread_id = 2;
    oneof payload {
        ClassDefinition class_def = 3;

        AllocationObject alloc_object = 4;
		AllocationArray alloc_array = 5;
        Deallocation dealloc = 6;

        MethodEnter method_enter = 7;

        GetField get_field = 8;
        PutFieldObject put_field_object = 9;
        PutFieldPrimitive put_field_primitive = 10;

        StoreObject store_object = 11;
        StorePrimitive store_primitive = 12;
        Load load = 13;
    }

}
